File: golden_search_method.m
clear;
clc;

tic
format long
% Initial values
a_lower = 1;  % lower limit
b_upper = 2;  % upper limit
ro = (3 - sqrt(5))/2; % constant step size (golden number)
eps = b_upper - a_lower;

% Assign initial value
a_old = a_lower;
b_old = b_upper;
iter = 0; % 

% Golden Section Algorithm
func = @(x) x^2 + 4*cos(x);  % you can change the 
while(eps > 10^-5)
   
    a_new = a_old + (b_old - a_old)*ro; % new lower limit
    b_new = b_old - (b_old - a_old)*ro; % new upper limit
    
    if (func(b_new) > func(a_new))
        b_old = b_new;       
    elseif (func(b_new) < func(a_new))
        a_old = a_new;
    end
    eps = b_old - a_old;
    iter = iter + 1;
    
end
timeElapsed = toc;

fprintf("\tGolden Search Method\n")
fprintf("Iteration : %d\nAverage computation time : %f\n",iter,timeElapsed);
fprintf("Minimum point of function : x = %f, f(x) = %f\n",a_new,func(a_new));

File: c++/golden_search_method.cpp
#include <bits/stdc++.h>

#define GR ((3 - sqrt(5))/2) // golden ratio

double func(double x){
    // Function to minimize, this can be changed to any function.
    // Parameters:
    //      :x: variable of function
    return x * x + 4 * cos(x);   
}


double goldensection(double aLower, double bUpper){
    // Implementation of golden section search method
    // Parameters:
    //      :aLower: lower limit
    //      :bUpper: upper limit

    int itr = 0;

    std::clock_t start;
    start = std::clock();
    
    double error = bUpper - aLower;
    double aPrev = aLower;
    double bPrev = bUpper;

    double aNew, bNew;
    while (error > exp10(-5)){
        aNew = aPrev + (bPrev - aPrev) * GR; // New lower limit
        bNew = bPrev - (bPrev - aPrev) * GR;

        if (func(bNew) > func(aNew)){
            bPrev = bNew; 
        }
        else if (func(bNew) < func(aNew)){
            aPrev = aNew;
        }

        error = fabs(bPrev - aPrev);
        itr++;
    }

    std::cout << "Golden Section Method" << std::endl;
    std::cout << "Iteration : " << itr << std::endl;
    std::cout << "Computation Time : "  << std::clock() - start << std::endl;
    std::cout << "Minimum point of function : " << "x = " << aNew  << ", f(x) = " << func(aNew) << std::endl; 
    
    return aNew;

}


int main(){


    goldensection(1,2);

    return 0;
}